README.md
# Gridlines for Tailwind CSS

An experimental default experience for Tailwind CSS that uses gridlines and fractions for spatial reasoning.

**Note:** This is under active development and may be subject to breaking changes until it reaches a stable version 1.0.0. Not recommended for production yet but please play around if this idea resonates with you.

## What is this?

Gridlines for Tailwind CSS is a reimagining of Tailwind's spacing system inspired by grid paper. It is a drop-in plugin that adds a preset which moves all spacing-related utilities over to to the [new system](https://github.com/ShiftLimits/tailwindcss-gridlines/wiki/Spacing). It also includes the [unit utility](https://github.com/ShiftLimits/tailwindcss-gridlines/wiki/Unit) to scale the spacing values per-element and some [overlay utilities](https://github.com/ShiftLimits/tailwindcss-gridlines/wiki/Gridlines) to help visualize the new spacing values.

The core idea is to switch from reasoning about layout in absolute pixels to reasoning in ratios and fractions relative to your base grid cell size. This allows for creating responsive designs that can be scaled by changing one value - `theme.base` - which is useful for adapting to device contexts where the user can be very close to the screen, such as tablets and mobile phones, or very far from the screen, such as so-called '10-foot' UI for TV and advertising panels.

## Installation

Install with your favorite NodeJS package manager:

```bash
$ npm install tailwindcss-gridlines

$ yarn add tailwindcss-gridlines
```

Add to your Tailwind CSS config file:

```js
// tailwind.config.js

module.exports = {
  theme: {
    // ...
  },
  plugins: [
    require('tailwindcss-gridlines')
  ],
}
```

## Documentation

Please see the [documentation Wiki](https://github.com/ShiftLimits/tailwindcss-gridlines/wiki) for a full description of what is new and what has changed.

## New Utilities

### Base Grid Cell Size

The core unit of gridlines is your base grid cell size. You can think of this as the equivalent of the 1cm in centimetre graph paper. Changing this value scales your layout proportionally.

| Name | Size |
| ---- | ---- |
| base | 24px |

#### Controlling Per-Element
- *See [unit documentation](https://github.com/ShiftLimits/tailwindcss-gridlines/wiki/Unit)*

The new `unit` utility can be used to adjust the relative sizing for an individual element and its children.

#### Customizing

You can customize the base unit by editing the `theme.base` value of your Tailwind config:

```js
// tailwind.config.js

module.exports = {
  theme: {
    base: '28px' // Changes the base grid cell size to 28px
  },
  // ...
}
```

You can make the base unit responsive using the breakpoint names defined in the `theme.screens` object as keys:

```js
// tailwind.config.js

module.exports = {
  theme: {
    base: {
      default: '20px', // Use a smaller grid cell size by default for smaller screens
      lg: '24px' // At the `lg` breakpoint, scale the grid size to 24px as these devices are probably held further away from the user
    }
  },
  // ...
}
```

### Gridlines
- *See [gridlines documentation](https://github.com/ShiftLimits/tailwindcss-gridlines/wiki/Gridlines)*

Overlay a grid to help visually aid the design process.

## Spacing Scale
- *See [spacing documentation](https://github.com/ShiftLimits/tailwindcss-gridlines/wiki/Spacing)*

Gridlines is a complete overhaul of the spacing scale in Tailwind CSS.

With vanilla Tailwind, the spacing scale uses decimals and is equivalent to a grid with a base unit size of `4px`. With Gridlines, the spacing scale uses fractions and you define the base unit size that the spacing scale is a fraction of. By default, Gridlines uses a base unit size of `24px`.

Example of the fraction scale from 0 to 1 with a base unit of `24px`:

| Name   | Size  | Pixels   |
| ------ | ----- | -------- |
| 0      | 0     | 0px      |
| px     | 1px   | 1px      |
| 1/8    | 0.125 | 3px      |
| 2/8    | 0.25  | 6px      |
| 3/8    | 0.375 | 9px      |
| 4/8    | 0.5   | 12px     |
| 5/8    | 0.625 | 15px     |
| 6/8    | 0.75  | 18px     |
| 7/8    | 0.875 | 21px     |
| 1/6    | 0.166 | 4px      |
| 2/6    | 0.333 | 7.992px  |
| 3/6    | 0.5   | 12px     |
| 4/6    | 0.666 | 16.008px |
| 5/6    | 0.833 | 20px     |
| 1/5    | 0.2   | 4.8px    |
| 2/5    | 0.4   | 9.6px    |
| 3/5    | 0.6   | 14.4px   |
| 4/5    | 0.8   | 19.12px  |
| 1/4    | 0.25  | 6px      |
| 2/4    | 0.5   | 12px     |
| 3/4    | 0.75  | 18px     |
| 1/3    | 0.333 | 7.992px  |
| 2/3    | 0.666 | 16.008px |
| 1/2    | 0.5   | 12px     |
| 1      | 1     | 24px     |

## Contributing

Feedback is welcome, please use GitHub issues appropriately. This idea has been rattling around in my head since 2018 so fresh perspectives may lead to greater insight.

If suggesting any major changes please make sure it is well reasoned and in line with the core principles of this project. Understand I am one busy person and require a description of the problem and a compelling argument for the proposed solution in order to consider it properly.

## License

Gridlines for Tailwind CSS is [MIT](LICENSE) licensed.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added
- Added support for the container queries provided by `@tailwindcss/container-queries`
- Added 6th denominations for spacing values between 0 and 5
- Added 5th denominations for spacing values between 1 and 5 (already present below 1)
- Added 3rd denominations for spacing values between 2 and 5 (already present below 2)

## [0.2.4] - 2022-02-11

### Added
- Added new range generator function to automatically generate a range of fraction utilities, accessible with `gridlines.generateRange` (see tests)
- Extended spacing scale to 60 and added half and quarter fractions for the whole scale

## [0.2.3] - 2022-01-09

### Added
- Add `unit` utility for adjusting the base grid cell size on a per-element basis

### Changed
- Increase `line-height` scale values to 5 units (from 2)
- Add Stroke Width
- Add outline and outline offset
- Fixed bottom positioning on gridlines utilities
- Gridlines utilities will no longer generate if `theme.gridlines` is `false`

## [0.2.2] - 2021-10-10
### Added
- Add `fit-content` utility for `width`, `min-width`, and `max-width`; and for `height`, `min-height`, and `max-height`
- Add `min-content` utility for `width`, `max-width`, `height`, `min-height`, and `max-height`
- Add `max-content` utility for `width`, `max-width`, `height`, `min-height`, and `max-height`


## [0.2.1] - 2021-09-24
### Added
- Add `CHANGELOG.md` following the Keep a Changelog format.

### Changed
- Capitalized `default` key to make sure default `borderWidth` utility is set appropriately (removes `border-default`)

## [0.2.0] - 2021-07-23
### Added
- New config variable in the theme object, `theme.base`, which controls the scale of your grid cells.
- New spacing system preset that is expressed as a fraction of the `theme.base` grid cell size.
- New `gridlines-` utilities that overlay a grid to help visualize the new spacing system.

src/index.ts
import plugin from 'tailwindcss/plugin'
import preflight from './plugins/preflight'
import gridlines from './plugins/gridlines'
import unit from './plugins/unit'
import { DEFAULT_BASE, FRACTIONS_1_TO_2, FRACTIONS_1_TO_5, PERCENTAGES, PERCENTAGES_FRACTIONS, UNIT_FRACTIONS, DEFAULT_RULE_COLOR, DEFAULT_RULE_SECONDARY_COLOR } from './constants';
import { generateFractionUtil, generateUnitMultipleString, generateUnitConfig, generateRange } from './utils';


module.exports = plugin(function(api:any) {
  const plugins = [preflight, gridlines, unit]
  for (let plugin of plugins) plugin(api)
}, {
  theme: {
    base: DEFAULT_BASE,
    unit: generateUnitConfig([
      '1/5',
      '2/5',
      '3/5',
      '4/5',
      '1/4',
      '2/4',
      '3/4',
      '1/3',
      '2/3',
      '1/2',
      '1',
      '1 1/8',
      '1 2/8',
      '1 3/8',
      '1 4/8',
      '1 5/8',
      '1 6/8',
      '1 7/8',
      '1 1/5',
      '1 2/5',
      '1 3/5',
      '1 4/5',
      '1 1/4',
      '1 2/4',
      '1 3/4',
      '1 1/3',
      '1 2/3',
      '1 1/2',
      '2',
      '2 1/4',
      '2 2/4',
      '2 3/4',
      '2 1/3',
      '2 2/3',
      '2 1/2',
      '3'
    ]),
    spacing: {
      px: '1px',
      '0': '0',
      ...UNIT_FRACTIONS,
      ...FRACTIONS_1_TO_5,
      ...generateRange(5, 60, { denominators: [2,4], inclusive: false })
    },
    borderRadius: {
      none: '0',
      ...UNIT_FRACTIONS,
      ...FRACTIONS_1_TO_5,
      full: '9999px',
    },
    borderWidth: {
      DEFAULT: '1px',
      '0': '0',
      ...UNIT_FRACTIONS
		},
    strokeWidth: {
        DEFAULT: '1px',
        '0': '0',
        ...UNIT_FRACTIONS
    },
    inset: (theme, { negative }) => ({
      auto: 'auto',
      ...theme('spacing'),
      ...negative(theme('spacing')),
      ...PERCENTAGES,
      ...PERCENTAGES_FRACTIONS,
      full: '100%',
      ...negative(PERCENTAGES),
      ...negative(PERCENTAGES_FRACTIONS),
      '-full': '-100%',
    }),
    lineHeight: {
      none: '1',
      tight: '1.25',
      snug: '1.375',
      normal: '1.5',
      relaxed: '1.625',
      loose: '2',
      ...UNIT_FRACTIONS,
      ...FRACTIONS_1_TO_5,
    },
    maxWidth: (theme, { breakpoints }) => ({
      none: 'none',
      ...theme('spacing'),
      ...theme('screens'),
      ...PERCENTAGES,
      ...PERCENTAGES_FRACTIONS,
      full: '100%',
      fit: 'fit-content',
      min: 'min-content',
      max: 'max-content',
      '1/4-screen': '25vw',
      '1/3-screen': '33vw',
      '1/2-screen': '50vw',
      '2/3-screen': '66vw',
      '3/4-screen': '75vw',
      screen: '100vw',
    }),
    minWidth: (theme) => ({
      '0': '0',
      ...theme('spacing'),
      ...theme('screens'),
      ...PERCENTAGES,
      ...PERCENTAGES_FRACTIONS,
      full: '100%',
      fit: 'fit-content',
      min: 'min-content',
      max: 'max-content',
      '1/4-screen': '25vw',
      '1/3-screen': '33vw',
      '1/2-screen': '50vw',
      '2/3-screen': '66vw',
      '3/4-screen': '75vw',
      screen: '100vw',
    }),
    width: theme => ({
      auto: 'auto',
      ...theme('spacing'),
      ...theme('screens'),
      ...PERCENTAGES,
      ...PERCENTAGES_FRACTIONS,
      full: '100%',
      fit: 'fit-content',
      min: 'min-content',
      max: 'max-content',
      '1/4-screen': '25vw',
      '1/3-screen': '33vw',
      '1/2-screen': '50vw',
      '2/3-screen': '66vw',
      '3/4-screen': '75vw',
      screen: '100vw',
    }),
    maxHeight: (theme) => ({
      none: 'none',
      ...theme('spacing'),
      ...PERCENTAGES,
      ...PERCENTAGES_FRACTIONS,
      full: '100%',
      fit: 'fit-content',
      min: 'min-content',
      max: 'max-content',
      '1/4-screen': '25vh',
      '1/3-screen': '33vh',
      '1/2-screen': '50vh',
      '2/3-screen': '66vh',
      '3/4-screen': '75vh',
      screen: '100vh',
    }),
    minHeight: (theme) => ({
      '0': '0',
      ...theme('spacing'),
      ...PERCENTAGES,
      ...PERCENTAGES_FRACTIONS,
      full: '100%',
      fit: 'fit-content',
      min: 'min-content',
      max: 'max-content',
      '1/4-screen': '25vh',
      '1/3-screen': '33vh',
      '1/2-screen': '50vh',
      '2/3-screen': '66vh',
      '3/4-screen': '75vh',
      screen: '100vh',
    }),
    height: theme => ({
      auto: 'auto',
      ...theme('spacing'),
      ...PERCENTAGES,
      ...PERCENTAGES_FRACTIONS,
      full: '100%',
      fit: 'fit-content',
      min: 'min-content',
      max: 'max-content',
      '1/4-screen': '25vh',
      '1/3-screen': '33vh',
      '1/2-screen': '50vh',
      '2/3-screen': '66vh',
      '3/4-screen': '75vh',
      screen: '100vh',
    }),
    gridlines: {
      ruleColor: DEFAULT_RULE_COLOR,
      ruleSecondaryColor: DEFAULT_RULE_SECONDARY_COLOR
    },
    ringOffsetWidth: {
      '0': '0',
      'px': '1px',
      ...UNIT_FRACTIONS
    },
    ringWidth: {
      DEFAULT: generateUnitMultipleString(0.125),
      '0': '0',
      'px': '1px',
      ...UNIT_FRACTIONS
    },
    outlineOffset: {
      '0': '0',
      'px': '1px',
      ...UNIT_FRACTIONS
    },
    outlineWidth: {
      '0': '0',
      'px': '1px',
      ...UNIT_FRACTIONS
    },
    textIndent: (theme, { negative }) => ({
      ...theme('lineHeight'),
      ...negative(theme('lineHeight'))
    }),
    translate: (theme, { negative }) => ({
      ...theme('spacing'),
      ...negative(theme('spacing')),
      ...PERCENTAGES,
      ...PERCENTAGES_FRACTIONS,
      full: '100%',
      ...negative(PERCENTAGES),
      ...negative(PERCENTAGES_FRACTIONS),
      '-full': '-100%',
    }),
    containers: (theme) => {
      let base = theme('base')
      if (typeof base == "object") base = base.default

      return {
        ...generateRange(1, 100, { denominators: [2, 4], inclusive: true, generateString(multiple) { return `${multiple * parseFloat(base)}px` } }),
        ...theme('screens'),
      }
    }
	},
  variants: {
    textIndent: ['responsive'],
  },
})

module.exports.generateRange = generateRange


src/utils.ts
export function generateUnitMultipleString(multiple:number) {
	return `calc(${multiple} * var(--twgl-current-base))`
}

type GenerateStringFn = typeof generateUnitMultipleString

import numericQuantity from 'numeric-quantity'
export function generateFractionUtil(fraction:string, generateString:GenerateStringFn = generateUnitMultipleString) {
	let decimal = numericQuantity(fraction)
	return { [fraction.replace(' ', '-')]: generateString(decimal) }
}

export function generateUnitConfig(fractions:string[]) {
	let config = {}
	for (let fraction of fractions) {
		let decimal = numericQuantity(fraction)
		Object.assign(config, { [fraction.replace(' ', '-')]: decimal })
	}

	return config
}

interface RangeOptions {
	denominators:number[]
	inclusive?:boolean
	generateString?:GenerateStringFn
}
export function generateRange(from:number, to:number, { denominators, inclusive, generateString }:RangeOptions = { denominators: [2, 4], inclusive: true }) {
	let range = {}

	for (let i = from; i <= to; i++) {
		if (i > 0 && (i != from || inclusive)) Object.assign(range, generateFractionUtil(`${i}`, generateString))
		if (i != to) {
		for (let denominator of denominators) {
				for (let numerator = 1; numerator < denominator; numerator++) {
					Object.assign(range, generateFractionUtil(`${i > 0 ? i+' ' : ''}${numerator}/${denominator}`, generateString))
				}
			}
		}
	}

	return range
}


src/plugins/gridlines.ts
import { DEFAULT_RULE_COLOR, DEFAULT_RULE_SECONDARY_COLOR, SVG_CELL_SIZE } from '../constants'

function stripSVGWhitespace(svg:string) {
  return svg.replace(/[\t\s\n]*(<.*>)[\t\s\n]*/g, '$1')
}

const DOT_SVG = (theme) => stripSVGWhitespace(`
  <svg viewbox='-0.5 -0.5 ${SVG_CELL_SIZE - 0.5} ${SVG_CELL_SIZE - 0.5}' xmlns='http://www.w3.org/2000/svg'>
    <rect shape-rendering='crispEdges' x='0' y='0' width='1' height='1' fill='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
  </svg>
`)
const UNIT_SVG = (theme) => stripSVGWhitespace(`
  <svg viewbox='-0.5 -0.5 ${SVG_CELL_SIZE - 0.5} ${SVG_CELL_SIZE - 0.5}' xmlns='http://www.w3.org/2000/svg'>
    <line shape-rendering='crispEdges' x1='0%' x2='100%' y1='-0.5' y2='-0.5' stroke-width='2' stroke='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
    <line shape-rendering='crispEdges' x1='0%' x2='100%' y1='100%' y2='100%' stroke-width='2' stroke='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
    <line shape-rendering='crispEdges' x1='-0.5' x2='-0.5' y1='0%' y2='100%' stroke-width='2' stroke='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
    <line shape-rendering='crispEdges' x1='100%' x2='100%' y1='0%' y2='100%' stroke-width='2' stroke='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
  </svg>
`)
const THIRDS_SVG = (theme) => stripSVGWhitespace(`
  <svg viewbox='-0.5 -0.5 ${SVG_CELL_SIZE - 0.5} ${SVG_CELL_SIZE - 0.5}' xmlns='http://www.w3.org/2000/svg'>
    <!-- Horizontal Lines -->
    <line shape-rendering='crispEdges' x1='0%' x2='100%' y1='-0.5' y2='-0.5' stroke-width='2' stroke='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
    <line shape-rendering='crispEdges' x1='0%' x2='100%' y1='100%' y2='100%' stroke-width='2' stroke='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
    <line shape-rendering='crispEdges' x1='-0.5' x2='100%' y1='33%' y2='33%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='4.125%' />
    <line shape-rendering='crispEdges' x1='-0.5' x2='100%' y1='66%' y2='66%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='4.125%' />

    <!-- Vertical Lines -->
    <line shape-rendering='crispEdges' x1='-0.5' x2='-0.5' y1='0%' y2='100%' stroke-width='2' stroke='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
    <line shape-rendering='crispEdges' x1='100%' x2='100%' y1='0%' y2='100%' stroke-width='2' stroke='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
    <line shape-rendering='crispEdges' x1='33%' x2='33%' y1='-0.5' y2='100%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='4.125%' />
    <line shape-rendering='crispEdges' x1='66%' x2='66%' y1='-0.5' y2='100%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='4.125%' />
  </svg>
`)
const QUADS_SVG = (theme) => stripSVGWhitespace(`
  <svg viewbox='-0.5 -0.5 ${SVG_CELL_SIZE - 0.5} ${SVG_CELL_SIZE - 0.5}' xmlns='http://www.w3.org/2000/svg'>
    <!-- Horizontal Lines -->
    <line shape-rendering='crispEdges' x1='0%' x2='100%' y1='-0.5' y2='-0.5' stroke-width='2' stroke='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
    <line shape-rendering='crispEdges' x1='0%' x2='100%' y1='100%' y2='100%' stroke-width='2' stroke='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
    <line shape-rendering='crispEdges' x1='-0.5' x2='100%' y1='25%' y2='25%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='4.125%' />
    <line shape-rendering='crispEdges' x1='-0.5' x2='100%' y1='50%' y2='50%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='4.125%' />
    <line shape-rendering='crispEdges' x1='-0.5' x2='100%' y1='75%' y2='75%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='4.125%' />

    <!-- Vertical Lines -->
    <line shape-rendering='crispEdges' x1='-0.5' x2='-0.5' y1='0%' y2='100%' stroke-width='2' stroke='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
    <line shape-rendering='crispEdges' x1='100%' x2='100%' y1='0%' y2='100%' stroke-width='2' stroke='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
    <line shape-rendering='crispEdges' x1='25%' x2='25%' y1='-0.5' y2='100%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='4.125%' />
    <line shape-rendering='crispEdges' x1='50%' x2='50%' y1='-0.5' y2='100%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='4.125%' />
    <line shape-rendering='crispEdges' x1='75%' x2='75%' y1='-0.5' y2='100%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='4.125%' />
  </svg>
`)
const FIFTHS_SVG = (theme) => stripSVGWhitespace(`
  <svg viewbox='-0.5 -0.5 ${SVG_CELL_SIZE - 0.5} ${SVG_CELL_SIZE - 0.5}' xmlns='http://www.w3.org/2000/svg'>
    <!-- Horizontal Lines -->
    <line shape-rendering='crispEdges' x1='0%' x2='100%' y1='-0.5' y2='-0.5' stroke-width='2' stroke='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
    <line shape-rendering='crispEdges' x1='0%' x2='100%' y1='100%' y2='100%' stroke-width='2' stroke='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
    <line shape-rendering='crispEdges' x1='-0.5' x2='100%' y1='20%' y2='20%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='5%' />
    <line shape-rendering='crispEdges' x1='-0.5' x2='100%' y1='40%' y2='40%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='5%' />
    <line shape-rendering='crispEdges' x1='-0.5' x2='100%' y1='60%' y2='60%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='5%' />
    <line shape-rendering='crispEdges' x1='-0.5' x2='100%' y1='80%' y2='80%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='5%' />

    <!-- Vertical Lines -->
    <line shape-rendering='crispEdges' x1='-0.5' x2='-0.5' y1='0%' y2='100%' stroke-width='2' stroke='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
    <line shape-rendering='crispEdges' x1='100%' x2='100%' y1='0%' y2='100%' stroke-width='2' stroke='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
    <line shape-rendering='crispEdges' x1='20%' x2='20%' y1='-0.5' y2='100%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='5%' />
    <line shape-rendering='crispEdges' x1='40%' x2='40%' y1='-0.5' y2='100%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='5%' />
    <line shape-rendering='crispEdges' x1='60%' x2='60%' y1='-0.5' y2='100%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='5%' />
    <line shape-rendering='crispEdges' x1='80%' x2='80%' y1='-0.5' y2='100%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='5%' />
  </svg>
`)
const OCTS_SVG = (theme) => stripSVGWhitespace(`
  <svg viewbox='-0.5 -0.5 ${SVG_CELL_SIZE - 0.5} ${SVG_CELL_SIZE - 0.5}' xmlns='http://www.w3.org/2000/svg'>
    <!-- Horizontal Lines -->
    <line shape-rendering='crispEdges' x1='0%' x2='100%' y1='-0.5' y2='-0.5' stroke-width='2' stroke='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
    <line shape-rendering='crispEdges' x1='0%' x2='100%' y1='100%' y2='100%' stroke-width='2' stroke='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
    <line shape-rendering='crispEdges' x1='-0.5' x2='100%' y1='12.5%' y2='12.5%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='3.125%'/>
    <line shape-rendering='crispEdges' x1='-0.5' x2='100%' y1='25%' y2='25%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='3.125%'/>
    <line shape-rendering='crispEdges' x1='-0.5' x2='100%' y1='37.5%' y2='37.5%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='3.125%'/>
    <line shape-rendering='crispEdges' x1='-0.5' x2='100%' y1='50%' y2='50%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='3.125%'/>
    <line shape-rendering='crispEdges' x1='-0.5' x2='100%' y1='62.5%' y2='62.5%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='3.125%'/>
    <line shape-rendering='crispEdges' x1='-0.5' x2='100%' y1='75%' y2='75%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='3.125%'/>
    <line shape-rendering='crispEdges' x1='-0.5' x2='100%' y1='87.5%' y2='87.5%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='3.125%'/>

    <!-- Vertical Lines -->
    <line shape-rendering='crispEdges' x1='-0.5' x2='-0.5' y1='0%' y2='100%' stroke-width='2' stroke='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
    <line shape-rendering='crispEdges' x1='100%' x2='100%' y1='0%' y2='100%' stroke-width='2' stroke='${theme('gridlines.ruleColor', DEFAULT_RULE_COLOR)}'/>
    <line shape-rendering='crispEdges' x1='12.5%' x2='12.5%' y1='-0.5' y2='100%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='3.125%'/>
    <line shape-rendering='crispEdges' x1='25%' x2='25%' y1='-0.5' y2='100%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='3.125%'/>
    <line shape-rendering='crispEdges' x1='37.5%' x2='37.5%' y1='-0.5' y2='100%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='3.125%'/>
    <line shape-rendering='crispEdges' x1='50%' x2='50%' y1='-0.5' y2='100%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='3.125%'/>
    <line shape-rendering='crispEdges' x1='62.5%' x2='62.5%' y1='-0.5' y2='100%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='3.125%'/>
    <line shape-rendering='crispEdges' x1='75%' x2='75%' y1='-0.5' y2='100%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='3.125%'/>
    <line shape-rendering='crispEdges' x1='87.5%' x2='87.5%' y1='-0.5' y2='100%' stroke-width='1' stroke='${theme('gridlines.ruleSecondaryColor', DEFAULT_RULE_SECONDARY_COLOR)}' stroke-dasharray='3.125%'/>
  </svg>
`)

export default function({ addUtilities, theme }:any) {
  if (!theme('gridlines')) return

  addUtilities(
    {
      '[class*="gridlines-"]::after': {
        'z-index': 1000000,
        content: `''`,
        display: 'block',
        position: 'absolute',
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,
        'background-size': 'var(--twgl-current-base) var(--twgl-current-base)',
        'pointer-events': 'none',
        'mix-blend-mode': 'difference'
      },
      '.gridlines-top-left::after': {
        'background-position': 'top left'
      },
      '.gridlines-left::after': {
        'background-position': 'left'
      },
      '.gridlines-bottom-left::after': {
        'background-position': 'bottom left'
      },
      '.gridlines-top-right::after': {
        'background-position': 'top right'
      },
      '.gridlines-right::after': {
        'background-position': 'right'
      },
      '.gridlines-bottom-right::after': {
        'background-position': 'bottom right'
      },
      '.gridlines-center::after': {
        'background-position': 'center'
      },
      '.gridlines-thirds::after': {
        'background-image': `url("data:image/svg+xml;utf8,${ THIRDS_SVG(theme) }")`
      },
      '.gridlines-quads::after': {
        'background-image': `url("data:image/svg+xml;utf8,${ QUADS_SVG(theme) }")`
      },
      '.gridlines-fifths::after': {
        'background-image': `url("data:image/svg+xml;utf8,${ FIFTHS_SVG(theme) }")`
      },
      '.gridlines-octs::after': {
        'background-image': `url("data:image/svg+xml;utf8,${ OCTS_SVG(theme) }")`
      },
      '.gridlines-dots::after': {
        'background-image': `url("data:image/svg+xml;utf8,${ DOT_SVG(theme) }")`
      },
      '.gridlines-unit::after': {
        'background-image': `url("data:image/svg+xml;utf8,${ UNIT_SVG(theme) }")`
      },
    }
  )
}
